securedrop/journalist_app/sessions.py

```diff
@@ -0,0 +1,278 @@
+from typing import Optional, Any, Dict, Tuple
+from flask.sessions import SessionInterface as FlaskSessionInterface
+from flask.sessions import SessionMixin, session_json_serializer
+from json.decoder import JSONDecodeError
+from redis import Redis
+from werkzeug.datastructures import CallbackDict
+from itsdangerous import URLSafeTimedSerializer, BadSignature
+
+from models import Journalist
+
+
+class ServerSideSession(CallbackDict, SessionMixin):
+    """Baseclass for server-side based sessions."""
+
+    def __init__(self, sid: str, token: str, lifetime: int = 0, initial: Any = None) -> None:
+        def on_update(self: ServerSideSession) -> None:
+            self.modified = True
+        if initial and 'uid' in initial:
+            self.set_uid(initial['uid'])
+            self.set_user()
+        else:
+            self.uid: Optional[int] = None
+            self.user = None
+        CallbackDict.__init__(self, initial, on_update)
+        self.sid = sid
+        self.token: str = token
+        self.lifetime = lifetime
+        self.is_api = False
+        self.to_destroy = False
+        self.to_regenerate = False
+        self.modified = False
+        self.flash: Optional[Tuple[str, str]] = None
+        self.locale: Optional[str] = None
+
+    def get_token(self) -> Optional[str]:
+        return self.token
+
+    def get_lifetime(self) -> datetime:
+        return datetime.now(timezone.utc) + timedelta(seconds=self.lifetime)
+
+    def set_user(self) -> None:
+        if self.uid is not None:
+            self.user = Journalist.query.get(self.uid)
+        if self.user is None:
+            # The uid has no match in the database, and this should really not happen
+            self.uid = None
+            self.to_destroy = True
+
+    def get_user(self) -> Optional[Journalist]:
+        return self.user
+
+    def get_uid(self) -> Optional[int]:
+        return self.uid
+
+    def set_uid(self, uid: int) -> None:
+        self.user = None
+        self.uid = uid
+
+    def logged_in(self) -> bool:
+        if self.uid is not None:
+            return True
+        else:
+            return False
+
+    def destroy(
+        self,
+        flash: Optional[Tuple[str, str]] = None,
+        locale: Optional[str] = None
+    ) -> None:
+        # The parameters are needed to pass the information to the new session
+        self.locale = locale
+        self.flash = flash
+        self.uid = None
+        self.user = None
+        self.to_destroy = True
+
+    def regenerate(self) -> None:
+        self.to_regenerate = True
+
+
+class SessionInterface(FlaskSessionInterface):
+
+    def _generate_sid(self) -> str:
+        return token_urlsafe(32)
+
+    def _get_signer(self, app: Flask) -> URLSafeTimedSerializer:
+        if not app.secret_key:
+            raise RuntimeError('No secret key set')
+        return URLSafeTimedSerializer(app.secret_key,
+                                      salt=self.salt)
```

Comment:
nit: the two methods above can be a `@staticmethod` as `self` is not used. It would also allow doing `SessionInterface._get_signer(app)` in the tests to re-use the implementation here.